src/out_references.rs

```diff
@@ -0,0 +1,495 @@
+///
+/// This type can be [trivially constructed][`crate::AsOut`] from:
+///
+///   - a `&'out mut [MaybeUninit<T>]` (main point of the type),
+///
+///   - a `&'out mut [T]` (to keep the ergonomics of being able to overwrite an
+///     already initialized value).
+///
+///       - To avoid "accidentally" leaking memory in this second case,
+///         either `T` must be `Copy` (sufficient condition to prove there is
+///         no drop glue), or you must first call
+///         [`.manually_drop_mut()`][`crate::ManuallyDropMut`]
+///         before the [`.as_out()`][`crate::AsOut`] "coercion".
+#[repr(transparent)]
+pub
+struct OutSlice<'out, T : 'out> (
+    &'out mut [MaybeUninit<T>],
+);
+
+impl<'out, T : 'out> From<&'out mut [T]> for OutSlice<'out, T>
+where
+    T : Copy,
+{
+    #[inline]
+    fn from (slice: &'out mut [T])
+      -> OutSlice<'out, T>
+    {
+        let len = slice.len();
+        OutSlice(unsafe {
+            // # Safety
+            //
+            //   - The API does not allow to write `MaybeUninit::uninit()` into
+            //     the pointee.
+            slice::from_raw_parts_mut(
+                slice.as_mut_ptr().cast(),
+                len,
+            )
+        })
+    }
+}
+
+impl<'out, T : 'out> From<&'out mut [ManuallyDrop<T>]> for OutSlice<'out, T> {
+    #[inline]
+    fn from (slice: &'out mut [ManuallyDrop<T>])
+      -> OutSlice<'out, T>
+    {
+        let len = slice.len();
+        OutSlice(unsafe {
+            // # Safety
+            //
+            //   - The API does not allow to write `MaybeUninit::uninit()` into
+            //     the pointee.
+            slice::from_raw_parts_mut(
+                slice.as_mut_ptr().cast(),
+                len,
+            )
+        })
+    }
+}
+
+impl<'out, T : 'out> From<&'out mut [MaybeUninit<T>]> for OutSlice<'out, T> {
+    #[inline]
+    fn from (slice: &'out mut [MaybeUninit<T>])
+      -> OutSlice<'out, T>
+    {
+        OutSlice(slice)
+    }
+}
+
+impl<'out, T : 'out> OutSlice<'out, T> {
+    #[inline]
+    pub
+    fn new (slice: impl Into<OutSlice<'out, T>>)
+      -> Self
+    {
+        slice.into()
+    }
+
+    #[inline]
+    pub
+    fn as_mut_ptr (self: &'_ mut Self)
+      -> *mut T
+    {
+        self.0.as_mut_ptr().cast()
+    }
+
+    #[inline]
+    pub
+    fn reborrow<'reborrow> (self: &'reborrow mut OutSlice<'out, T>)
+      -> OutSlice<'reborrow, T>
+    where
+        'out : 'reborrow,
+    {
+        OutSlice(&mut *self.0)
+    }
+
+    /// Shorthand for [`.reborrow()`][`OutSlice::reborrow`]
+    #[inline]
+    pub
+    fn r<'reborrow> (self: &'reborrow mut OutSlice<'out, T>)
+      -> OutSlice<'reborrow, T>
+    where
+        'out : 'reborrow,
+    {
+        self.reborrow()
+    }
+
+    #[inline]
+    pub
+    fn idx<Index> (self: OutSlice<'out, T>, idx: Index)
+      -> Option<Index::Output>
+    where
+        Index : SliceIndex<'out, T>,
```

Comment:
You're right.

`.get()` would shadow the `Deref`'s `.get()`, so I'd rather `get_mut()`. I have ended up going  with `.get_out()` since it yields an out reference.